streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java

```diff
@@ -49,48 +51,209 @@ public StreamTask restoringTaskFor(final TopicPartition partition) {
     List<StreamTask> allTasks() {
         final List<StreamTask> tasks = super.allTasks();
         tasks.addAll(restoring.values());
+        tasks.addAll(suspended.values());
         return tasks;
     }
 
     @Override
     Set<TaskId> allAssignedTaskIds() {
         final Set<TaskId> taskIds = super.allAssignedTaskIds();
         taskIds.addAll(restoring.keySet());
+        taskIds.addAll(suspended.keySet());
         return taskIds;
     }
 
     @Override
     boolean allTasksRunning() {
-        return super.allTasksRunning() && restoring.isEmpty();
+        return super.allTasksRunning() && restoring.isEmpty() && suspended.isEmpty();
     }
 
-    RuntimeException closeAllRestoringTasks() {
+    @Override
+    void closeTask(final StreamTask task, final boolean clean) {
+        if (suspended.containsKey(task.id())) {
+            task.closeSuspended(clean, false, null);
+        } else {
+            task.close(clean, false);
+        }
+    }
+
+    Set<TaskId> suspendedTaskIds() {
+        return suspended.keySet();
+    }
+
+    RuntimeException suspend(final Set<TaskId> revokedTasks, final List<TopicPartition> revokedChangelogs) {
+        final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
+
+        log.trace("Suspending running {} {}", taskTypeName, runningTaskIds());
+        firstException.compareAndSet(null, suspendTasks(revokedTasks, revokedChangelogs));
+
+        log.trace("Close created {} {}", taskTypeName, created.keySet());
+        firstException.compareAndSet(null, closeNonRunningTasks(revokedTasks, revokedChangelogs));
+
+        return firstException.get();
+    }
+
+    private RuntimeException suspendTasks(final Set<TaskId> tasks, final List<TopicPartition> revokedChangelogs) {
+        final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
+        for (final Iterator<TaskId> it = tasks.iterator(); it.hasNext(); ) {
+            final TaskId id = it.next();
+            if (!running.containsKey(id) || suspended.containsKey(id)) {
```

Comment:
Feel better to put entire block in `running.containsKey(id) && suspended.containsKey(id)` than continue, but up to you.